server/src/main/java/com/linecorp/centraldogma/server/storage/encryption/DefaultEncryptionStorageManager.java

```diff
@@ -0,0 +1,418 @@
+        requireNonNull(projectName, "projectName");
+        requireNonNull(repoName, "repoName");
+        requireNonNull(wdek, "wdek");
+        final byte[] wdekKeyBytes = repoWdekDbKey(projectName, repoName).getBytes(StandardCharsets.UTF_8);
+        try {
+            final byte[] existingWdek = rocksDb.get(wdekKeyBytes);
+            if (existingWdek != null) {
+                throw new EncryptionStorageException(
+                        "WDEK of " + projectRepo(projectName, repoName) + " already exists");
+            }
+        } catch (RocksDBException e) {
+            throw new EncryptionStorageException(
+                    "Failed to check existence of WDEK for " + projectRepo(projectName, repoName), e);
+        }
+
+        // RocksDB is thread-safe.
+        try {
+            rocksDb.put(wdekKeyBytes, wdek);
+        } catch (RocksDBException e) {
+            throw new EncryptionStorageException(
+                    "Failed to store WDEK of " + projectRepo(projectName, repoName), e);
+        }
+    }
+
+    private static String projectRepo(String projectName, String repoName) {
+        return projectName + '/' + repoName;
+    }
+
+    private static String repoWdekDbKey(String projectName, String repoName) {
+        return "wdeks/" + projectName + '/' + repoName;
+    }
+
+    @Override
+    public void removeWdek(String projectName, String repoName) {
+        requireNonNull(projectName, "projectName");
+        requireNonNull(repoName, "repoName");
+        try {
+            rocksDb.delete(repoWdekDbKey(projectName, repoName).getBytes(StandardCharsets.UTF_8));
+        } catch (RocksDBException e) {
+            throw new EncryptionStorageException(
+                    "Failed to remove WDEK of " + projectRepo(projectName, repoName), e);
+        }
+    }
+
+    @Override
+    public byte[] get(byte[] key) {
+        requireNonNull(key, "key");
+        try {
+            return rocksDb.get(key);
+        } catch (RocksDBException e) {
+            throw new EncryptionStorageException("Failed to get value", e);
+        }
+    }
+
+    @Override
+    public byte[] getMetadata(byte[] metadataKey) {
+        requireNonNull(metadataKey, "metadataKey");
+        try {
+            return rocksDb.get(metadataColumnFamilyHandle, metadataKey);
+        } catch (RocksDBException e) {
+            throw new EncryptionStorageException("Failed to get metadata. key: " +
+                                                 new String(metadataKey, StandardCharsets.UTF_8), e);
+        }
+    }
+
+    @Override
+    public void put(byte[] metadataKey, byte[] metadataValue, byte[] key, byte[] value) {
+        try (WriteBatch writeBatch = new WriteBatch();
+             WriteOptions writeOptions = new WriteOptions()) {
+            writeBatch.put(metadataColumnFamilyHandle, metadataKey, metadataValue);
+            writeBatch.put(defaultColumnFamilyHandle, key, value);
+            rocksDb.write(writeOptions, writeBatch);
```

Comment:
Question) Just to make sure I understood correctly, will the write also be flushed immediately?

I'm assuming the following:

1. Assuming CD data needs to be replicated, we don't have to worry about the WAL.
2. If the java application is uncleanly terminated (e.g. hypervisor maintenance), we do have to worry about the WAL - either by replicating based on a cleanly terminated cluster, or somehow moving the WAL together with data files.